File: julia/2022/day4.jl
using Test

function part_one(input)
    ranges = split.(split(input, "\n"), ",")
    overlaps = 0
    for r in ranges
        first = parse.(Int64, split(r[1], "-"))
        second = parse.(Int64, split(r[2], "-"))

        if (first[1] >= second[1] && first[2] <= second[2]) ||
           (second[1] >= first[1] && second[2] <= first[2])
            overlaps += 1
        end
    end
    return overlaps
end

function part_two(input)
    ranges = split.(split(input, "\n"), ",")
    overlaps = 0
    for r in ranges
        first = parse.(Int64, split(r[1], "-"))
        second = parse.(Int64, split(r[2], "-"))

        if second[1] <= first[1] <= second[2] ||
           second[1] <= first[2] <= second[2] ||
           first[1] <= second[1] <= first[2]
            overlaps += 1
        end
    end
    return overlaps
end

function test()
    example = """2-4,6-8
2-3,4-5
5-7,7-9
2-8,3-7
6-6,4-6
2-6,4-8"""

    @test part_one(example) == 2
    @test part_two(example) == 4
end

File: ruby/2023/day4.rb
require './commons/assert'
require 'set'

def get_matches(cards)
  cards.map do |c|
    winning, mine = c.split(' | ').map { |ns| Set[*ns.split(' ').map(&:to_i)] }
    winning.intersection(mine).length
  end
end

def part_one(input)
  cards = input.split("\n").map { |c| c.split(': ')[1] }
  get_matches(cards).map { |m| m.zero? ? 0 : 2**(m - 1) }.sum
end

def part_two(input)
  cards = input.split("\n").map { |c| c.split(': ')[1] }
  matches = get_matches(cards).map { |c| [c, 1] } # idx -> matches, amount
  matches.each_with_index do |p, idx|
    ((idx + 1)..(idx + p[0])).each { |ni| matches[ni][1] += p[1] }
  end
  matches.map { |p| p[1] }.sum
end

def test
  example = %(Card 1: 41 48 83 86 17 | 83 86  6 31 17  9 48 53
Card 2: 13 32 20 16 61 | 61 30 68 82 17 32 24 19
Card 3:  1 21 53 59 44 | 69 82 63 72 16 21 14  1
Card 4: 41 92 73 84 69 | 59 84 76 51 58  5 54 83
Card 5: 87 83 26 28 32 | 88 30 70 12 93 22 82 36
Card 6: 31 18 13 56 72 | 74 77 10 23 35 67 36 11)

  assert part_one(example), 13
  assert part_two(example), 30
  0
end
